scoring/tests/test_scoring.py

```diff
@@ -0,0 +1,244 @@
+            scorer.validate(None)
+
+    def test_invalid_robot_token_characters(self):
+        scorer = self.construct_scorer(
+            {'ABC': 'X'},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    def test_lower_case_zone_token_characters(self):
+        self.zone_tokens[0] = 's'
+        scorer = self.construct_scorer(
+            {},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    def test_lower_case_robot_token_characters(self):
+        scorer = self.construct_scorer(
+            {'ABC': 's'},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    def test_invalid_zone_token_characters_in_zone_without_robot(self):
+        self.zone_tokens[3] += 'X'
+        scorer = self.construct_scorer(
+            {},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    def test_lower_case_zone_token_characters_in_zone_without_robot(self):
+        self.zone_tokens[3] += 's'
+        scorer = self.construct_scorer(
+            {},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    # Missing tokens
+
+    def test_less_than_26_tokens_seen(self):
+        self.zone_tokens[0] = 'S'
+        scorer = self.construct_scorer(
+            {},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    def test_less_than_26_tokens_seen_in_zone_without_robot(self):
+        self.zone_tokens[3] = 'S'
+        scorer = self.construct_scorer(
+            {},
+            self.zone_tokens,
+        )
+        with self.assertRaises(InvalidScoresheetException):
+            scorer.validate(None)
+
+    # Tolerable input deviances
+
+    def test_space_in_zone_tokens(self):
+        self.zone_tokens[0] = shuffled(self.zone_tokens[0] + "      ")
+        self.assertScores(
+            {'ABC': 81, 'DEF': 81},
+            {},
+            self.zone_tokens,
+        )
+
+    def test_space_in_robot_tokens(self):
+        self.zone_tokens[0] = 'B' * 4 + 'S' * 2
+        self.assertScores(
+            {'ABC': 51, 'DEF': 81},
+            {'ABC': ' B  S G '},
+            self.zone_tokens,
+        )
+
+    def test_left_scoring_zone_not_specified(self):
+        self.teams_data = {
+            'ABC': {'zone': 0, 'present': True},
+            'DEF': {'zone': 1, 'present': True},
+        }
+        self.assertScores(
+            {'ABC': 81, 'DEF': 81},
+            {},
+            self.zone_tokens,
+        )
+
+    def test_left_scoring_zone_but_absent(self):
```

Comment:
Does this need its own section? Not really a tolerable input deviance

Reply:
Ah, good spot.

Reply:
8f80987